[JIT] Throw error when no_grad is used as a decorator

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50129 [JIT] Throw error when no_grad is used as a decorator**

**Summary**
JIT supports `torch.jit.no_grad` as a context manager, but not as a
decorator. When it is used as the latter, it compiles successfully
without any errors, but there are no IR nodes emitted to implement
`no_grad` functionality.

This commit adds code in `get_jit_def` that throws an
`torch.jit.frontend.NotSupportedError` when a function decorated with
`@torch.no_grad` is encountered.

**Test Plan**
This commit adds a unit test to `test_with.py` (since that's where the
other `no_grad` tests are) to check that an error is thrown when
scripting functions decorated with `@torch.no_grad`.

**Fixes**
This commit fixes #44768.

Differential Revision: [D25855125](https://our.internmc.facebook.com/intern/diff/D25855125)

Commit:
[JIT] Throw error when no_grad is used as a decorator

**Summary**
JIT supports `torch.jit.no_grad` as a context manager, but not as a
decorator. When it is used as the latter, it compiles successfully
without any errors, but there are no IR nodes emitted to implement
`no_grad` functionality.

This commit adds code in `get_jit_def` that throws an
`torch.jit.frontend.NotSupportedError` when a function decorated with
`@torch.no_grad` is encountered.

**Test Plan**
This commit adds a unit test to `test_with.py` (since that's where the
other `no_grad` tests are) to check that an error is thrown when
scripting functions decorated with `@torch.no_grad`.

**Fixes**
This commit fixes #44768.

[ghstack-poisoned]

Commit:
Update on "[JIT] Throw error when no_grad is used as a decorator"


**Summary**
JIT supports `torch.jit.no_grad` as a context manager, but not as a
decorator. When it is used as the latter, it compiles successfully
without any errors, but there are no IR nodes emitted to implement
`no_grad` functionality.

This commit adds code in `get_jit_def` that throws an
`torch.jit.frontend.NotSupportedError` when a function decorated with
`@torch.no_grad` is encountered.

**Test Plan**
This commit adds a unit test to `test_with.py` (since that's where the
other `no_grad` tests are) to check that an error is thrown when
scripting functions decorated with `@torch.no_grad`.

**Fixes**
This commit fixes #44768.

[ghstack-poisoned]

Commit:
Update on "[JIT] Throw error when no_grad is used as a decorator"


**Summary**
JIT supports `torch.jit.no_grad` as a context manager, but not as a
decorator. When it is used as the latter, it compiles successfully
without any errors, but there are no IR nodes emitted to implement
`no_grad` functionality.

This commit adds code in `get_jit_def` that throws an
`torch.jit.frontend.NotSupportedError` when a function decorated with
`@torch.no_grad` is encountered.

**Test Plan**
This commit adds a unit test to `test_with.py` (since that's where the
other `no_grad` tests are) to check that an error is thrown when
scripting functions decorated with `@torch.no_grad`.

**Fixes**
This commit fixes #44768.

[ghstack-poisoned]

Commit:
Update on "[JIT] Throw error when no_grad is used as a decorator"


**Summary**
JIT supports `torch.jit.no_grad` as a context manager, but not as a
decorator. When it is used as the latter, it compiles successfully
without any errors, but there are no IR nodes emitted to implement
`no_grad` functionality.

This commit adds code in `get_jit_def` that throws an
`torch.jit.frontend.NotSupportedError` when a function decorated with
`@torch.no_grad` is encountered.

**Test Plan**
This commit adds a unit test to `test_with.py` (since that's where the
other `no_grad` tests are) to check that an error is thrown when
scripting functions decorated with `@torch.no_grad`.

**Fixes**
This commit fixes #44768.

[ghstack-poisoned]

Commit:
Update on "[JIT] Throw error when no_grad is used as a decorator"


**Summary**
JIT supports `torch.jit.no_grad` as a context manager, but not as a
decorator. When it is used as the latter, it compiles successfully
without any errors, but there are no IR nodes emitted to implement
`no_grad` functionality.

This commit adds code in `get_jit_def` that throws an
`torch.jit.frontend.NotSupportedError` when a function decorated with
`@torch.no_grad` is encountered.

**Test Plan**
This commit adds a unit test to `test_with.py` (since that's where the
other `no_grad` tests are) to check that an error is thrown when
scripting functions decorated with `@torch.no_grad`.

**Fixes**
This commit fixes #44768.

[ghstack-poisoned]

gmagogsfm commented on torch/jit/frontend.py:
```diff
@@ -217,6 +217,35 @@ def remove_prefix(text, prefix):
+    for decorator in fn_def.decorator_list:
+        if isinstance(decorator, ast.Call):
+            decorator_name = get_decorator_name(decorator)
+            if decorator_name == "torch.no_grad":
```
Would this be safe against things like import aliasing?

import torch as to
from torch import no_grad

If it is not safe, could we fetch object according to name and compare against actual `torch.no_grad` object?

SplitInfinity replied:
Good point, I'm so used to seeing `@torch.no_grad` everywhere that I didn't realize. Yes, I think it is possible to fetch the object and compare against `torch.no_grad`.

SplitInfinity replied:
Done.

Commit:
Update on "[JIT] Throw error when no_grad is used as a decorator"


**Summary**
JIT supports `torch.jit.no_grad` as a context manager, but not as a
decorator. When it is used as the latter, it compiles successfully
without any errors, but there are no IR nodes emitted to implement
`no_grad` functionality.

This commit adds code in `get_jit_def` that throws an
`torch.jit.frontend.NotSupportedError` when a function decorated with
`@torch.no_grad` is encountered.

**Test Plan**
This commit adds a unit test to `test_with.py` (since that's where the
other `no_grad` tests are) to check that an error is thrown when
scripting functions decorated with `@torch.no_grad`.

**Fixes**
This commit fixes #44768.

[ghstack-poisoned]

gmagogsfm reviewed (approved):
Looks good! Thanks for making this change.

I do have a high level question, though. `check_fn_decorators_supported` is phrased to check whether all decorators present are supported. And it is currently using a blocklist-based approach (only `torch.no_grad` is excluded). Does this mean all other decorators are supported? Should we consider an allowlist-based approach instead if that's not the case?

SplitInfinity commented:
> Looks good! Thanks for making this change.
> 
> I do have a high level question, though. `check_fn_decorators_supported` is phrased to check whether all decorators present are supported. And it is currently using a blocklist-based approach (only `torch.no_grad` is excluded). Does this mean all other decorators are supported? Should we consider an allowlist-based approach instead if that's not the case?

A decorator itself can be user-defined or be part of a `torch`, and it's possible that a user's decorator is harmless as far as scripting is concerned. Except for `@ignore` and `@unused`, decorators in general are completely ignored by TorchScript, and Python will have "evaluated" (not sure if I'm using the right word here) them before scripting. `no_grad` is also special because it is used both as a context manager and a decorator; this is only true for `dual_level` and `enable_grad` (neither of which we support, so it might be worth adding them to this blocklist).

Commit:
Update on "[JIT] Throw error when no_grad is used as a decorator"


**Summary**
JIT supports `torch.jit.no_grad` as a context manager, but not as a
decorator. When it is used as the latter, it compiles successfully
without any errors, but there are no IR nodes emitted to implement
`no_grad` functionality.

This commit adds code in `get_jit_def` that throws an
`torch.jit.frontend.NotSupportedError` when a function decorated with
`@torch.no_grad` is encountered.

**Test Plan**
This commit adds a unit test to `test_with.py` (since that's where the
other `no_grad` tests are) to check that an error is thrown when
scripting functions decorated with `@torch.no_grad`.

**Fixes**
This commit fixes #44768.

[ghstack-poisoned]
